velox/expression/tests/SimpleFunctionTest.cpp

```diff
@@ -551,6 +552,109 @@ TEST_F(SimpleFunctionTest, nonDefaultNullBehavior) {
+
+    return true;
+  }
+};
+
+TEST_F(SimpleFunctionTest, defaultContainsNullsBehavior) {
+  registerFunction<DefaultContainsNullsBehaviorFunction, bool, Array<int32_t>>(
+      {"default_contains_nulls_behavior"});
+
+  // Make a vector with some null arrays, and some arrays containing nulls.
+  auto arrayVector = makeVectorWithNullArrays<int32_t>(
+      {std::nullopt, {{1, std::nullopt}}, {{std::nullopt, 2}}, {{1, 2, 3}}});
+
+  // Check that default contains nulls behavior functions don't get called on a
+  // null input.
+  auto result = evaluate<SimpleVector<bool>>(
+      "default_contains_nulls_behavior(c0)", makeRowVector({arrayVector}));
+  auto expected = makeNullableFlatVector<bool>(
+      {std::nullopt, std::nullopt, std::nullopt, true});
+  assertEqualVectors(expected, result);
+}
+
+// Test that function with non-default contains nulls behavior.
+// Returns 0 for null and input[0] * 2 if call is invoked.
+// Returns input[0] * 8 if callNullFree is invoked.
+template <typename T>
+struct NonDefaultContainsNullsBehaviorFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool callNullable(
+      out_type<int32_t>& out,
+      const arg_type<Array<int32_t>>* input) {
+    if (input == nullptr) {
+      out = 0;
+    } else {
+      out = (*input)[0].value() * 2;
+    }
+
+    return true;
+  }
+
+  FOLLY_ALWAYS_INLINE bool callNullFree(
+      out_type<int32_t>& out,
+      const null_free_arg_type<Array<int32_t>>& input) {
+    out = input[0] * 8;
+
+    return true;
+  }
+};
+
+TEST_F(SimpleFunctionTest, nonDefaultContainsNullsBehavior) {
+  registerFunction<
+      NonDefaultContainsNullsBehaviorFunction,
+      int32_t,
+      Array<int32_t>>({"non_default_contains_nulls_behavior"});
+
+  // Make a vector with a NULL.
+  auto arrayVectorWithNull = makeVectorWithNullArrays<int32_t>(
+      {std::nullopt, {{1, 2}}, {{3, 2}}, {{2, 2, 3}}});
+  // Make a vector with an array that contains NULL.
+  auto arrayVectorContainsNull = makeVectorWithNullArrays<int32_t>(
+      {{{4, std::nullopt}}, {{1, 2}}, {{3, 2}}, {{2, 2, 3}}});
+  // Make a vector that's NULL-free.
+  auto arrayVectorNullFree =
+      makeArrayVector<int32_t>({{4, 5}, {1, 2}, {3, 2}, {2, 2, 3}});
+
+  // CallNullable should be invoked so we should see the value in the first
+  // position multiplied by 2 or 0 if it's NULL.
+  auto resultWithNull = evaluate<FlatVector<int32_t>>(
+      "non_default_contains_nulls_behavior(c0)",
+      makeRowVector({arrayVectorWithNull}));
+  auto expectedWithNull = makeFlatVector<int32_t>({0, 2, 6, 4});
+  assertEqualVectors(expectedWithNull, resultWithNull);
+
+  // CallNullable should be invoked so we should see the value in the first
+  // position multiplied by 2 or 0 if it's NULL.
+  auto resultContainsNull = evaluate<FlatVector<int32_t>>(
+      "non_default_contains_nulls_behavior(c0)",
+      makeRowVector({arrayVectorContainsNull}));
+  auto expectedContainsNull = makeFlatVector<int32_t>({8, 2, 6, 4});
```

Comment:
shall we add a test for variadic<row<map, array>>
this would cover all the readers. and would catch the current issue.

Reply:
Will do!